File: Codelearn/Sasuke 36/kthSpecialNumber.cpp
string kthSpecialNumber(std::vector<int> arr, int k)
{
    sort(arr.begin(), arr.end());
    int len = arr.size(), i = 1, x, y;
    long s = len;
    if(arr[0] == 0){
    	x = len - 1;
    	y = 1;
    } else {
    	x = 1;
    	y = 0;
    }
	while(s < k){
		++i;
		s += x * pow(len, i - y);
	}
	s -= x * pow(len, i - y);
	x = k - s - 1;
	int n, a[11], l = 0;
	bool check = false;
	std::vector<string> v;
	for(int j = 1; j <= i; ++j){
		a[j] = arr[0];
	}
	while(!check){
		string s = "";
		for(int j = 1; j <= i; j++){
  			s += (char)(a[j] + 48);
 		}
 		v.push_back(s);
 		++l;
		int m = i;
		while(a[m] == arr[len - 1] && m > 0){
			m--;
		}
		if(l == x + 1)
			check = true;
		else{
			int index = distance(arr.begin(), find(arr.begin(), arr.end(), a[m]));
			a[m] = arr[index + 1];
			for(int p = m + 1; p <= i; p++){
				a[p] = arr[0];
			}
		}
	}
	return v[x];
}

File: Codelearn/Traning/kthSpecialNumber.js
let a;
let current;
let map;

function kthSpecialNumber(arr, k) {
   if(arr[0] == 0 && arr.length == 1) return "-1";

   arr.sort((a,b) => a-b);

   let size = 0,
      count = 0;
   while (k > count) {
      k -= count;
      size++;
      if(arr[0] == 0)
      {
         count = (arr.length - 1)*Math.pow(arr.length, size - 1);
      } else count = Math.pow(arr.length, size);
   }

   map = Array(size).fill(null);
   a = k;

   find(0, arr);

   return current;
}

function find(index, arr) {
   if(a <= 0) return ;
   if(index >= map.length)
   {
      a--;
      if(a == 0) current = map.join('');
   } else {
      for(let i = 0; i < arr.length; i++){
         if(a <= 0) break;
         if(index == 0 && arr[i] == 0) continue;
         map[index] = arr[i];
         find(index + 1, arr);
      }
   }
}